### ISSUE TITLE
Error propagates with `return x()` but not with `return try x()` inside recursion

### ISSUE THREAD

**Issue by wooster0** [ziglang/zig#15669]
### Zig Version

0.11.0-dev.3031+f40539e5d

### Steps to Reproduce and Observed Behavior

So this might be entirely expected behavior but it still struck me as a bit odd because so far I took `return try x()` as being pretty much the same as `return x()`.
```zig
const Value = union(enum) {
    values: []const Value,
    b,
};

fn x(value: Value) !void {
    switch (value.values[0]) {
        .values => return x(value.values[0]),
        .b => return error.a,
    }
}

test {
    const a = Value{
        .values = &[1]Value{
            .{
                .values = &[1]Value{.{ .b = {} }},
            },
        },
    };
    _ = try x(a);
}
```
```
$ zig test x.zig
Test [1/1] test_0... FAIL (a)
/home/mooster/Desktop/os/x.zig:9:15: 0x20f068 in x (test)
        .b => return error.a,
              ^
```
But now change `return x(value.values[0]),` inside `x()` to  `return try x(value.values[0]),`:
```
$ zig test x.zig
All 1 tests passed.
```
This is how in practice in my program for a while errors were just being ignored and I was confused. This might just be perfectly expected behavior though so I don't know how much can be done about this. Maybe `return try` could be disallowed and you should write `return`? I expect them to behave the same generally but it's a bit different here with recursion apparently.

### Expected Behavior

For the input, `x()` always returns `error.a` and the error should be properly propagated up to the caller so that `Test [1/1] test_0... FAIL (a)` appears when testing the code, regardless of whether I use `return` or `return try`.

**Comment by rohlem** [comment]
To me that looks 100% like a bug.
I've further (on version `0.11.0-dev.2828+5b9e528bc`) narrowed it down to involve the deduced error set:
When the return type of `x` is changed to `error{a}!void` it reproduces the same when using `return try`.

Also note: reordering the switch cases similarly makes the error re-appear.
So it really seems the compiler is incorrectly deducing an empty error set at the point where x calls itself,
and marks it as "fully resolved" ignoring the rest of the function body.

P.S.: I also write all of my code this way, and think it is indeed better style because the `try` explicitly signals the option of an error (would be a compile error if it wasn't an error union).
Using just `return` a reader might overlook that possibility for error.

**Comment by andrewrk** [comment]
Fix reverted in ad391ad399b26cf6a781e01c6daecf48bb764478.

**Comment by r00ster91** [comment]
This bug report can now be seen as a proposal as well. The main motivation here is that the current behavior is rather confusing.
And as @rohlem mentioned just reordering switch cases here can make behavior change which is unarguably a bug.
I wonder what would be an alternative to the behavior I proposed in Expected Behavior?

**Comment by nektro** [comment]
I would even potentially label this a miscompilation. if the return type of `x` is an error union then `return try x()` not being the same as `return x()` is a massive break in contract of `try y` being `y catch |err| return err`. this should hold regardless of whether or not it's in a recursion context or not, no?

**Comment by rohlem** [comment]
As far as I understand, everything written in this issue here is indeed a bug report of a miscompilation. (Please someone point out if this actually isn't the case.)

What was reverted was the particular fix implemented in https://github.com/ziglang/zig/commit/5aa9628de3c6637f45b9d8cf8cbd19c422a74f6f.
While I don't fully understand the internals enough to explain the changed semantics, a previous test named "ret_ptr doesn't cause own inferred error set to be resolved" was removed in that commit.
I assume the test was removed because the new behavior broke it, and that is the actual change of behavior that caused Andrew to revert the change, and would need to be proposed first.

Therefore I believe the current status is we're still open to / trying to find a solution to this current issue that retains the behavior for this old test case as well.

**Comment by andrewrk** [comment]
I have a happy little patch that passes the new test case:

```diff
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -30732,18 +30732,10 @@ fn analyzeIsNonErrComptimeOnly(
                     else => return .none,
                 },
             }
-            for (ies.inferred_error_sets.keys()) |other_ies_index| {
-                if (set_ty == other_ies_index) continue;
-                const other_resolved =
-                    try sema.resolveInferredErrorSet(block, src, other_ies_index);
-                if (other_resolved == .anyerror_type) {
-                    ies.resolved = .anyerror_type;
-                    return .none;
-                }
-                if (ip.indexToKey(other_resolved).error_set_type.names.len != 0)
-                    return .none;
-            }
-            return .bool_true;
+            // We do not have a comptime answer because this inferred error
+            // set is not resolved, and an instruction later in this function
+            // body may or may not cause an error to be added to this set.
+            return .none;
         },
         else => switch (ip.indexToKey(set_ty)) {
             .error_set_type => |error_set_type| {
@@ -30771,18 +30763,10 @@ fn analyzeIsNonErrComptimeOnly(
                                 else => return .none,
                             },
                         }
-                        for (ies.inferred_error_sets.keys()) |other_ies_index| {
-                            if (set_ty == other_ies_index) continue;
-                            const other_resolved =
-                                try sema.resolveInferredErrorSet(block, src, other_ies_index);
-                            if (other_resolved == .anyerror_type) {
-                                ies.resolved = .anyerror_type;
-                                return .none;
-                            }
-                            if (ip.indexToKey(other_resolved).error_set_type.names.len != 0)
-                                return .none;
-                        }
-                        return .bool_true;
+                        // We do not have a comptime answer because this inferred error
+                        // set is not resolved, and an instruction later in this function
+                        // body may or may not cause an error to be added to this set.
+                        return .none;
                     }
                 }
                 const resolved_ty = try sema.resolveInferredErrorSet(block, src, set_ty);
```

Unfortunately it is still regressing the other test case that I don't want to regress:

```zig
fn doTheTest() !void {
    errdefer @compileError("bad");

    @breakpoint();
    return try @This().foo();
}

fn foo() !void {}
```

However, if we look at this ZIR:

```
        %105 = field_call(.auto, %102, "foo", []) node_offset:22:16 to :22:29
        %106 = try(%105, {
          %107 = err_union_code(%105) node_offset:22:12 to :22:29
          %108 = defer({
            %91 = dbg_stmt(2, 14)
            %92 = str("bad")
            %93 = as_node(@InternPool.Index.slice_const_u8_type, %92) node_offset:19:28 to :19:33
            %94 = compile_error(%93) node_offset:19:14 to :19:34
            %95 = break_inline(%0, @InternPool.Index.void_value)
          })
          %109 = dbg_stmt(5, 12)
          %110 = ret_node(%107) node_offset:22:12 to :22:29
        }) node_offset:22:12 to :22:29
        %111 = store_node(%99, %106) node_offset:22:12 to :22:29
```

It actually does succeed at not triggering the compile error in the `try`.

It's actually the `return` that is not successfully noticing that it is impossible for there to be an error in this case:

```
        %112 = load(%99) node_offset:22:5 to :22:29
        %113 = ret_is_non_err(%112) node_offset:22:5 to :22:29
        %114 = condbr(%113, {
          %115 = restore_err_ret_index(%4294967211, %4294967211)
          %116 = dbg_stmt(5, 5)
          %117 = ret_load(%99) node_offset:22:5 to :22:29
        }, {
          %118 = defer({
            %91 = dbg_stmt(2, 14)
            %92 = str("bad")
            %93 = as_node(@InternPool.Index.slice_const_u8_type, %92) node_offset:19:28 to :19:33
            %94 = compile_error(%93) node_offset:19:14 to :19:34
            %95 = break_inline(%0, @InternPool.Index.void_value)
          })
          %119 = dbg_stmt(5, 5)
          %120 = ret_load(%99) node_offset:22:5 to :22:29
        }) node_offset:22:5 to :22:29
```

*This* is the compile error that is being triggered, and it is not correct to edit `analyzeIsNonErrComptimeOnly` to fix it. So I am confident that reverting 5aa9628de3c6637f45b9d8cf8cbd19c422a74f6f was the right move.

Two ideas to solve this:
* The compiler could to be taught that the result of `try` cannot be an error. Perhaps `return try` could be lowered with a special ZIR instruction since it's a fairly common idiom.
* That still leaves `return @This().foo();` to solve, which also emits the unwanted compile error. Perhaps AstGen can emit a different string of ZIR instructions that do not pointlessly store/load the return value and instead check is_non_err directly on the result of the function call.

In conclusion, I am confident that solving this bug does not require changing the language. It's a nice trick to use `@compileError` to forbid failure beyond a certain point in the function.
